### PR TITLE
add ons in string directory - Boyer_Moore_Search

### PR DESCRIPTION
**Pull request by Dharni0607**
#697, I have added Boyer Moore search function(bad character heuristic). Please review and let me know if any changes are required. Thanks. ;)

### CODE REVIEW

**Comment by Erfaniaa** [comment]
I think it's "Boyer" Moore Search. If I'm right, please let me change it.

**Comment by Dharni0607** [comment]
Yes, it's my bad. I will edit it. Thanks.

File: strings/Boyer_Moore_Search.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+"""
+This algorithm tries to find the pattern in given text using Bad Character Heuristic method. 
+The bad-character rule considers the character in Text at which there is a mis-match. The next occurrence of that character to the left in Pattern is found, and a shift which brings that occurrence in line with the mismatched occurrence in Text is proposed. If the mismatched character does not occur to the left in Pattern, a shift is proposed that moves the entirety of Pattern past the point of mismatch
+
+Complexity : O(n/m)
+    n=length of main string
+    m=length of pattern string
+"""
+
+class BayerMooreSearch:
```
Comment (Erfaniaa) [inline strings/Boyer_Moore_Search.py]: Please change the dictation mistake in the class name.

File: strings/Boyer_Moore_Search.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    """ following functions tries to find the mis-matched character's index in text from last, 
+    If all characters are matched, -1 will be returned"""
+    def misMatchInText(self, currentPos):
+        for i in range(self.patLen-1, -1, -1):
+            if self.pattern[i] != self.text[currentPos + i]:
+                return currentPos + i
+        return -1
+        
+    def badCharacterHeuristic(self):
```
Comment (Erfaniaa) [inline strings/Boyer_Moore_Search.py]: Please use snake_case for all methods.
Like: bad_character_heuristic

File: strings/Boyer_Moore_Search.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        positions = []
+        for i in range(self.textLen - self.patLen + 1):
+            misMatchIndex = self.misMatchInText(i)
```
Comment (Erfaniaa) [inline strings/Boyer_Moore_Search.py]: Use snake_case: mismatch_index
Try to use it in all variable names.

File: strings/Boyer_Moore_Search.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """ following functions tries to find the index of given char in pattern, 
+    if not found -1 will be returned"""
+    def matchInPattern(self, char):
+        for i in range(self.patLen-1, -1, -1):
+            if char == self.pattern[i]:
+                return i
+        return -1
+
+
+    """ following functions tries to find the mis-matched character's index in text from last,
```
Comment (Erfaniaa) [inline strings/Boyer_Moore_Search.py]: Place docstrings after the "def" line.
Explain the parameters and the return value separately.

Please Google "Python docstrings conventions".
https://www.datacamp.com/community/tutorials/docstrings-python#second-head

File: strings/Boyer_Moore_Search.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                i = misMatchIndex - matchIndex   #shifting
+        return positions
+
+
+if __name__ == '__main__': 
+    text = "ABAAABCDBBABCDDEBCABC"
+    pattern = "ABC" 
+    bms = BayerMooreSearch(text, pattern)
+    positions = bms.badCharacterHeuristic()
+    if(len(positions) == 0):
```
Comment (Erfaniaa) [inline strings/Boyer_Moore_Search.py]: Unnecessary parentheses

File: strings/Boyer_Moore_Search.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        print("No match found")
+    else:
+        print("Pattern found in following positions: ")
+        print(positions)
```
Comment (Erfaniaa) [inline strings/Boyer_Moore_Search.py]: I ran your code and the algorithm seems to work fine.

**Pull request merged**